lib/common_test/src/ct_suite.erl

```diff
@@ -0,0 +1,132 @@
+              ct_status/0,
+              ct_groups_def/0,
+              ct_group_props/0,
+              ct_testcase_ref/0,
+              ct_group_ref/0,
+              ct_subgroups_def/0,
+              ct_tests_def/0,
+              ct_info/0
+             ]).
+
+-type ct_testname() :: atom().
+-type ct_groupname() :: atom().
+-type ct_config() :: [{Key :: atom(), Value :: term()}].
+-type ct_status() :: ok |
+            skipped |
+            failed.
+-type ct_group_props() :: [
+                parallel |
+                sequence |
+                shuffle |
+                {shuffle, Seed :: {integer(), integer(), integer()}} |
+                {ct_group_repeat_type(), ct_test_repeat()}
+            ] |
+            default.
+-type ct_group_repeat_type() :: repeat |
+            repeat_until_all_ok |
+            repeat_until_all_fail |
+            repeat_until_any_ok |
+            repeat_until_any_fail.
+-type ct_test_repeat() :: integer() |
+            forever.
+-type ct_groups_def() :: {ct_groupname(), ct_group_props(), [
+                ct_testname() |
+                ct_groups_def() |
+                {group, ct_groupname()} |
+                ct_testcase_ref()
+            ]}.
+-type ct_subgroups_def() :: {ct_groupname(), ct_group_props()} |
+            {ct_groupname(), ct_group_props(), ct_subgroups_def()}.
+-type ct_group_ref() :: {group, ct_groupname()} |
+            {group, ct_groupname(), ct_group_props()} |
+            {group, ct_groupname(), ct_group_props(), ct_subgroups_def()}.
+-type ct_testcase_ref() :: {testcase, ct_testname(), ct_testcase_repeat_props()}.
+-type ct_testcase_repeat_props() :: {repeat, ct_test_repeat()} |
+            {repeat_until_ok, ct_test_repeat()} |
+            {repeat_until_fail, ct_test_repeat()}.
+-type ct_info() :: {timetrap, ct_info_timetrap()} |
+            {require, ct_info_required()} |
+            {require, Name :: atom(), ct_info_required()} |
+            {userdata, UserData :: term()} |
+            {silent_connections, Conns :: [atom()]} |
+            {stylesheet, CSSFile :: string()} |
+            {ct_hooks, CTHs :: ct_hooks()}.
+-type ct_info_timetrap() :: MilliSec :: integer() |
+            {seconds, integer()} |
+            {minutes, integer()} |
+            {hours, integer()} |
+            {Mod :: module(), Func :: atom(), Args :: list()} |
+            ct_info_timetrap_fun().
+-type ct_info_timetrap_fun() :: fun().
+-type ct_info_required() :: Key :: atom() |
+            {Key :: atom(), SubKeys :: ct_info_required_subkeys()} |
+            {Key :: atom(), Subkey :: atom()} |
+            {Key :: atom(), Subkey :: atom(), SubKeys :: ct_info_required_subkeys()}.
+-type ct_info_required_subkeys() :: SubKey :: atom() |
+            [SubKey :: atom()].
+-type ct_hooks() :: [
+                CTHModule :: atom() |
+                {CTHModule :: atom(), CTHInitArgs :: term()} |
+                {CTHModule :: atom(), CTHInitArgs :: term(), CTHPriority :: term()}
```

Comment:
[`common_test Module:group/1`](https://erlang.org/doc/man/common_test.html#Module:group-1) refers to `CTHPriority` but this is not specified. I'm tentatively specifying it as `term()`.

Also, the doc. itself doesn't refer to any "priority", and I've never used this element either.

The same goes for [`common_test Module:suite/0`](https://erlang.org/doc/man/common_test.html#Module:suite-0).

Reply:
suite/0 can be used to specify 'ct_hooks', 'timetrap', 'userdata' and 'require' configuration for a suite according to the documentation (common_test).
group/1 can override the values for suite/0 for a special group.  Example from base64_SUITE.erl in stdlib 
```

suite() ->
    [{ct_hooks,[ts_install_cth]},
     {timetrap,{minutes,4}}].

all() ->
    [base64_encode, base64_decode, base64_otp_5635,
     base64_otp_6279, big, illegal, mime_decode, mime_decode_to_string,
     {group, roundtrip}].

groups() ->
    [{roundtrip, [parallel],
      [roundtrip_1, roundtrip_2, roundtrip_3, roundtrip_4]}].

group(roundtrip) ->
    %% valgrind needs a lot of time
    [{timetrap,{minutes,10}}].

```
CTHPriority is an integer, better  specified in the ct_hooks documentation (both manual page and UG for common_test) It defines in which order hooks should be run.

Reply:
fe68aa0